### ISSUE TITLE
Figure out auto-scaling of Doctor

### ISSUE THREAD

**Issue by mlissner** [freelawproject/courtlistener#2103]
Right now I'm noticing that doctor is scaling pretty poorly. Lots of pods, most using very little CPU, but some using a ton of CPU. Simultaneously, lots of dropped connections to Doctor. 

I think there are a couple fixes to this:

1. We are currently using four workers per pod. We should switch this to one.
2. Inside the pod, we tesseract uses four threads. We should disable this.

**Comment by mlissner** [comment]
Our retries and moving to a single worker might be helping here, errors are down. 

If that doesn't work, the next options are to:

1. Set up the kube-proxy to [be ipvs mode](https://medium.com/@selfieblue/how-to-enable-ipvs-mode-on-aws-eks-7159ec676965) ([more info here](https://jicowan.medium.com/the-problem-with-kube-proxy-enabling-ipvs-on-eks-169ac22e237e)), and to remove the kube-proxy add-on, replacing it with one that is self-managed via a manifest ([here's why](https://github.com/aws/containers-roadmap/issues/142#issuecomment-1096812905)). 

2. Stop using ClusterIP service for the deployment and instead switch to AWS's Network Load Balancer, which can use connection count for routing HTTP target groups.

Both of those solutions look hard, fragile, and annoying, so let's hope we don't go there.

**Comment by mlissner** [comment]
@albertisfu can you chime in about whether we're still seeing errors with connections to doctor, please?

**Comment by albertisfu** [comment]
Sure, seems there are still a few errors related to connections to doctor on sentry, eg:
https://sentry.io/organizations/freelawproject/issues/3348541548/?project=5257254&query=is%3Aunresolved+ConnectionError&statsPeriod=14d

https://sentry.io/organizations/freelawproject/issues/3321009155/?project=5257254&query=is%3Aunresolved+ConnectionError&statsPeriod=14d

These tasks currently have a max of 3 retries, might be necessary to increase a bit?  maybe 5?
 
There are a few more from `cl_scrape_opinions` like:
https://sentry.io/organizations/freelawproject/issues/3306639559/?project=5257254&query=is%3Aunresolved+ConnectionError&statsPeriod=14d

There is no celery task involved here, so we might need to retry the doctor connection if it fails, testing the `@retry` decorator to use here, there is a problem because on the retry no request parameters are sent to doctor and returns a bad response error, I'll look for an alternative to retry doctor calls outside of a celery task.

**Comment by mlissner** [comment]
Actually, @albertisfu says the latest changes haven't been deployed, so let's check in here again after those are deployed.

**Comment by mlissner** [comment]
@albertisfu can you check again and give an opinion about whether this seems fixed to you?

**Comment by albertisfu** [comment]
@mlissner Sure! I've done a review of these related errors on sentry, here are my findings:

At a first glance, there are fewer errors since the latest changes on retries but they continue to happen.

e.g:
[process_recap_pdf task calls extract_recap_pdf task](https://sentry.io/organizations/freelawproject/issues/3348541548/events/latest/?project=5257254&query=is%3Aunresolved+ConnectionError&statsPeriod=14d):

I checked some of these errors and they are  RecapDocument related:
[207857453](https://www.courtlistener.com/docket/63218527/40/vondra-v-city-of-billings/): Content was extracted successfully.
[29702374](https://www.courtlistener.com/docket/7004943/108/united-states-v-madrid-paz/): This one is weird, the extraction succeeded but seems that the content returned doesn't match the actual content. Is this normal?
[7587652](https://www.courtlistener.com/docket/4363309/53/the-huntington-national-bank-v-taylor-building-products-inc/): Same case as above.
[188582390](https://www.courtlistener.com/docket/60375544/20/creda-software-inc-adversary-proceeding/): Content was extracted successfully.
[205064106](https://www.courtlistener.com/docket/63221137/14/16/park-consulting-group-v-marengo-township/): Content was extracted successfully.


[204816133](https://www.courtlistener.com/docket/12998617/1/watson-v-kalamazoo-regional-educational-services-agency/): Failed to extract content.
[208018878](https://www.courtlistener.com/docket/63270022/29/trinity-tank-car-inc-v-davis-frost-inc/): Failed to extract content.
[207805428](https://www.courtlistener.com/docket/61688506/72/georgia-state-conference-of-the-naacp-v-state-of-georgia/):  Failed to extract content.
[208189239](https://www.courtlistener.com/docket/63588594/11/reges-v-cauce/):  Failed to extract content.

So, half of these errors are actual failures and half are just retries that logged the error initially but they succeeded.

Errors related to: [extract_doc_content](https://sentry.io/organizations/freelawproject/issues/3321009155/?project=5257254&query=is%3Aunresolved+ConnectionError&statsPeriod=14d)

[7797026](https://www.courtlistener.com/opinion/7853160/insight-public-sector-inc-v-united-states/)
[7796887](https://www.courtlistener.com/opinion/7853021/in-re-morrow-park-holding-llc/)
[7622550](https://www.courtlistener.com/opinion/7687713/sanofi-aventis-us-v-mylan/)
[7374904](https://www.courtlistener.com/opinion/7454051/in-re-marriage-of-conopeotis/)
[6500877](https://www.courtlistener.com/opinion/6624519/state-of-tennessee-v-jamie-l-tice/)

All of those errors are opinions that failed to extract their content.
`extract_doc_content` is a task that is called directly (not within another task) so I think that's the reason that here we only see actual failures. Seems that the task ran out of retries (5)
One thing in common with these documents is that they are large documents with more than 70 pages.


There are some other errors like this related to [process_recap_docket_history_report](https://sentry.io/organizations/freelawproject/issues/3424271736/?project=5257254&query=is%3Aunresolved+ConnectionError&statsPeriod=14d) 

The couple of errors we have here:
[208055979](https://www.courtlistener.com/docket/63346805/8/estate-of-carmen-e-crespo-v-democratic-peoples-republic-of-korea/)
[206228491](https://www.courtlistener.com/docket/60033914/240/interstate-underground-warehouse-and-industrial-pa/)

Belong to documents that were processed properly by Doctor. 
So seems that the problem here is also a task being called within another task, so retries are logged to sentry.


Finally, we had this error:
[test_for_meta_redirections](https://sentry.io/organizations/freelawproject/issues/3464469862/?project=5257254&query=is%3Aunresolved+ConnectionError&statsPeriod=14d)

In this function, we are using the retry decorator because this is not a task. Seems that the decorator is working according to sentry.
So the problem here seems it ran out of retries.

In conclusion:

1.- Some tasks that log these errors have the same problem that `PacerLoginException`, a task that is called a function within another task. So retries are logged on sentry even though the task succeeded.

_Here we could apply a solution like the one proposed to  `fetch_attachment_page` which is to convert these functions that are called within another task and also wrap these functions within a new task to use in places where are called directly._

2.- There are errors for documents that failed after increasing retries. So yes it could mean that there is some problem with k8s config for doctor?

**Comment by mlissner** [comment]
> [29702374](https://www.courtlistener.com/docket/7004943/108/united-states-v-madrid-paz/): This one is weird, the extraction succeeded but seems that the content returned doesn't match the actual content. Is this normal?

That's due to https://github.com/freelawproject/doctor/issues/139

> Here we could apply a solution like the one proposed to fetch_attachment_page which is to convert these functions that are called within another task and also wrap these functions within a new task to use in places where are called directly.

Great.

>  One thing in common with these documents is that they are large documents with more than 70 pages.

Could be that these are timing out? Maybe doctor needs more time? 

>  So the problem here seems it ran out of retries.

Could this one be a circular redirect?

**Comment by albertisfu** [comment]
Perfect. I'll check the fix to avoid sentry logs within retries on these tasks.

About `document-extract` service, it has a timeout of 15 minutes. I'll do some tests locally to check how much time it takes to extract a 70 pages document.

I'll check the `test_for_meta_redirections` logs and check if it might be a circular redirect.

**Comment by albertisfu** [comment]
@mlissner  Seems that I've found the problem triggering the `ConnectionError` on sentry for `extract_doc_content` task.

These errors belong to opinions:
https://sentry.io/organizations/freelawproject/issues/3481420732/events/oldest/?project=5257254&query=is%3Aunresolved+ConnectionError&statsPeriod=14d

Specifically, the problem happens when a document with many pages is sent to doctor and the document contains at least one image (the small stamp at the top), e.g:
[state_of_tennessee_v._zachary_rye_adams.pdf](https://github.com/freelawproject/courtlistener/files/9788201/state_of_tennessee_v._zachary_rye_adams.pdf)

When sending opinions to Doctor seems that the `ocr_available` param is always set to True.

So Doctor will try to extract by OCR, the `ocr_needed` method detects that the document has images and returns that ocr_needed is True. So the process of extracting by OCR continues. Debugging seems that the process gets stuck in the method `rasterize_pdf`. 

So the microservice call times out (after 15 minutes), the task is retried 5 times, every time it fails and finally the error is logged to sentry.

Extracting that same document without OCR is fast.

I think the solution here from CL side is to use a similar approach that the one we use for recap documents where first we try to extract the text without OCR and then we can check if we get some content, if not then we can try to extract it by OCR. 

This problem also made me think if this also could happen to large documents that really need OCR. Might we need to increase the time out? Might be useful to use a different timeout when extracting by OCR?

Let me know what you think and I'll work on the solution.

**Comment by mlissner** [comment]
All suggestions make sense to me. Thank you for your analysis here.

**Comment by mlissner** [comment]
Alberto, when you get back to this one, can you please summarize what you think is still needed and we can prioritize accordingly?

**Comment by albertisfu** [comment]
I've looked for `ConnectionError`, `ConnectTimeout`,  or `ReadTimeout` on Sentry in the last 30 days and I couldn't find any of these errors related to Doctor requests. So seems the last fixes addressed on #2332 are working.

I found some `ConnectionError` on sentry but related to requests to external court resources from `cl_scrape_opinions` should we open an issue to analyze and propose solutions to these errors?

**Comment by mlissner** [comment]
Lots of the scrapers are in bad shape at the moment. We'll get to them eventually, but not just yet. Closing this!
